src/main/java/org/opengis/cite/trainingdmlai10part2/trainingdataset/TrainingDatasetClassTests.java

```diff
@@ -100,44 +76,226 @@ public void verifyDocumentIsJSON() {
+                    nodes.add(node.get(array));
+                }
+            }
+
+            for (JsonNode targetNode : nodes) {
+                for (int i = 0; i < targetNode.size(); i++) {
+                    JsonNode currentNode = targetNode.get(i);
+                    String nodeClass = currentNode.getClass().toString();
+                    if (nodeClass.endsWith("com.fasterxml.jackson.databind.node.ObjectNode")) {
+                        Set<ValidationMessage> errors = schema.validate(currentNode);
+                        Iterator it = errors.iterator();
+                        while (it.hasNext()) {
+                            sb.append("Item " + i + " has error " + it.next() + ".\n");
+                        }
+                    }
+                }
+            }
+//Data type and values should validate in json schema
+//            List<JsonNode> nodesAtLeastOne = new ArrayList<JsonNode>();
+//
+//            for (String array : arrayToFetchAtLeastOne) {
+//                if (node.has(array)) {
+//                    nodesAtLeastOne.add(node.get(array));
+//                }
+//            }
+//
+//            for (JsonNode targetNode : nodesAtLeastOne) {
+//                if (targetNode.size() > 1) Assert.fail(String.format("Node %s has more than one item.", targetNode));
+//
+//                for (int i = 0; i < targetNode.size(); i++) {
+//                    JsonNode currentNode = targetNode.get(i);
+//                    String nodeClass = currentNode.getClass().toString();
+//                    if (nodeClass.endsWith("com.fasterxml.jackson.databind.node.ObjectNode")) {
+//                        Set<ValidationMessage> errors = schema.validate(currentNode);
+//                        Iterator it = errors.iterator();
+//                        while (it.hasNext()) {
+//                            sb.append("Item " + i + " has error " + it.next() + ".\n");
+//                        }
+//                    }
+//                }
+//            }
 
+        } catch (Exception e) {
+            sb.append(e.getMessage());
+            e.printStackTrace();
+        }
     }
 
     /**
-     * Checks the behavior of the trim function.
+     * Verify that JSON instance documents claiming conformance to this specification contain valid URL values according to Uniform Resource Identifier (URI): Generic Syntax <a href="https://datatracker.ietf.org/doc/html/rfc3986#section-4.1">RFC 3986 Section 4.1</a>. A URL value can be absolute or relative and may have an optional fragment identifier.
      */
     @Test(description = "Implements Abstract Test 4 (/conf/base/jsonbasetype/url)")
     public void VerifyThatURLsAreValid() {
-        throw new SkipException("Not implemented yet.");
+        if (!testSubject.isFile()) {
+            Assert.assertTrue(testSubject.isFile(), "No file selected. ");
+        }
 
-    }
+        String URI_REGEX = "^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?";
```

Comment:
How was the regex generated?
It seems that URLs like `"dataURL": "s3://pixalytics-ogc-api/TrainingDML-AI/input/image/tile05_rgb.tif"` are not accepted.